### PR TITLE
Add hooks for register_buffer/module/parameter

### LINKED ISSUE #85837: Parameter Registration Hook
### 🚀 The feature, motivation and pitch

As far as I can tell, there is no hook that can be run as a parameter is registered. This could just be due to the fact that it was never needed before.

The motivation for wanting this feature now would be to wrap every single parameter that is being registered in a function call that could read/modify its state. Without this hook, we would have to wrap each and every parameter initialization manually. This makes things potentially harder to read and more difficult to maintain versus a hook that can be easily be enabled/disabled. Moreover, its not even possible to do this for parameters that are defined in modules like `torch.nn.Linear` where we have no control over the implementation.

An example of what this proposal could look like is to have something like the following in `module.py`:
```python
...
_global_parameter_registration_hooks: Dict[int, Callable] = OrderedDict()

def register_module_parameter_registration_hook(hook: Callable[..., None]) -> RemovableHandle:
    ...
...

class Module:
    ...
    def register_parameter(self, name: str, param: Optional[Parameter]) -> None:
        ...
        for func in _global_parameter_registration_hooks:
            param = func(param)
        self._parameters[name] = param
    ...
...
```
Also, while we're at it, we can also do this for `register_buffer` and `add_module`.

Would there be any reason this hook could not or should not be added? 

### Alternatives

Note, the wrapping function we want to hook to the parameter registration is not something that can be called after all the parameters are registered by retrieving them via `model.parameters()`. It needs to be called on a per parameter basis before the next parameter is registered

### Additional context

_No response_

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### PR DESCRIPTION
**Pull request by antoniojkim**
As described in the issue, this PR adds hooks to be run when `register_parameter`, `register_buffer` and `register_module` are called.

Fixes #85837

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345 @saketh-are

### CODE REVIEW

**Comment by albanD** [comment]
Hi @antoniojkim thanks for sending a PR for this!
- we review changes to nn.Module very carefully and in particular changing the direct setting of _modules to a register_module is something we should make sure is safe.
- could you split the formatting change in a separate PR? That would make it a lot easier to review this one.

**Comment by antoniojkim** [comment]
>     * could you split the formatting change in a separate PR? That would make it a lot easier to review this one.

Would this PR pass the linting check in CI without the proper formatting?

**Comment by albanD** [comment]
> Would this PR pass the linting check in CI without the proper formatting?

No. But we can merge the formatting PR quickly and so it will disappear from this PR.

**Comment by antoniojkim** [comment]
I've undone all the formatting changes

> changing the direct setting of _modules to a register_module is something we should make sure is safe.

Turns out that the behaviour is slightly different. So, I've reverted it back to how it was before. There is now multiple copies of the code that calls the buffer registration and module registration hooks

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -90,6 +95,78 @@ def __setstate__(self, state: Dict):
 _EXTRA_STATE_KEY_SUFFIX = '_extra_state'
 
 
+def register_buffer_registration_hook(hook: Callable[..., None]) -> RemovableHandle:
```
Comment (albanD) [inline torch/nn/modules/module.py:98]: Could you name these with a "module" in the name to match the other hooks?

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -90,6 +95,78 @@ def __setstate__(self, state: Dict):
...
+    r"""Registers a buffer registration hook common to all modules.
+
+    .. warning ::
+
+        This adds global state to the `nn.Module` module
+
+    The hook will be called every time :func:`register_buffer` is invoked.
+    It should have the following signature::
+
+        hook(name, module) -> None or modified input
```
Comment (albanD) [inline torch/nn/modules/module.py:108]: Shouldn't the buffer be passed in?
Also for consistency, could you pass the module first: `hook(module, name, buffer) -> None or new buffer`

Comment (albanD) [reply torch/nn/modules/module.py:108]: Same for the others below

**Review by albanD — COMMENTED** [review]
Looks pretty good!
We would need a couple tests and update to the signature and that will be good to go!

File: torch/nn/modules/module.py (hunk 2 above)
Comment (antoniojkim) [reply torch/nn/modules/module.py:108]: yep, good catch. That's my bad for just copying the docstring over and not replacing all instances 😅

File: torch/nn/modules/module.py [hunk 3]
```diff
@@ -90,6 +96,78 @@ def __setstate__(self, state: Dict):
...
+    r"""Registers a buffer registration hook common to all modules.
+
+    .. warning ::
+
+        This adds global state to the `nn.Module` module
+
+    The hook will be called every time :func:`register_buffer` is invoked.
+    It should have the following signature::
+
+        hook(name, buffer) -> None or modified input
```
Comment (albanD) [inline torch/nn/modules/module.py:109]: This should still be `hook(module, name, buffer) -> None or new buffer`

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -90,6 +96,78 @@ def __setstate__(self, state: Dict):
...
+    r"""Registers a module registration hook common to all modules.
+
+    .. warning ::
+
+        This adds global state to the `nn.Module` module
+
+    The hook will be called every time :func:`register_module` is invoked.
+    It should have the following signature::
+
+        hook(name, module) -> None or modified input
```
Comment (albanD) [inline torch/nn/modules/module.py:133]: I think this one should be `hook(module, name, submodule) -> None or new submodule`

File: torch/nn/modules/module.py [hunk 5]
```diff
@@ -90,6 +96,78 @@ def __setstate__(self, state: Dict):
...
+    r"""Registers a parameter registration hook common to all modules.
+
+    .. warning ::
+
+        This adds global state to the `nn.Module` module
+
+    The hook will be called every time :func:`register_parameter` is invoked.
+    It should have the following signature::
+
+        hook(name, param) -> None or modified input
```
Comment (albanD) [inline torch/nn/modules/module.py:157]: This one should be `hook(module, name, param) -> None or new param`

**Review by albanD — COMMENTED** [review]
Sorry for the misunderstanding. I think we can do a small update for the signatures.
The calls will need to be updated as well to pass in the right arguments to the hooks.

**Review by albanD — COMMENTED** [review]
Signatures look good!
I think we only need two things now:
- Update the call sites to pass in the right arguments
- Add some basic tests for each of these hooks

**Review by albanD — APPROVED** [review]
SGTM!
Thanks for taking the time to add full testing!

**Pull request closed without merging**